ApolloParseException/NPE with multiple fragment references

Using Apollo: `0.3.3-SNAPSHOT`

I have the following `.graphl` files:

Search.graphql

```
query SearchGraphQL($term: String, $limit: Int) {
  search(term: $term, limit: $limit) {
      results {
      id
      type
      display_type
      content {
        id
        type
        name
        link
        sort_order
        ...OfferFragment
        ...BonusFragment
      }
    }
  }
}
```

OfferFragment.graphql

```
fragment OfferFragment on Offer {
  id
  activated
  amount
}
```

BonusFragment.graphql

```
fragment BonusFragment on Bonus {
  id
  type
  name
}
```

The response from the server looks like this:

```
{
  "data": {
    "search": {
      "results": [
        {
          "id": "-100",
          "type": "ContentModule",
          "display_type": "rows",
          "content": [
            {
              "id": "105995",
              "type": "Offer",
              "name": "20% cash back & $2.00 Bonus at Hallmark Gold Crown",
              "link": "retailer/any/offer/105995",
              "sort_order": null,
              "activated": false,
              "amount": 0
            }
          ]
        },
        {
          "id": "-120",
          "type": "ContentModule",
          "display_type": "rows",
          "content": [
            {
              "id": "2768",
              "type": "Bonus",
              "name": "Boxed Dash for Cash Bonus",
              "link": "bonuses/bonus/2768",
              "sort_order": null
            },
            {
              "id": "2780",
              "type": "Bonus",
              "name": "Stoli Vodka Bonus",
              "link": "bonuses/bonus/2780",
              "sort_order": null
            }
          ]
        }
      ]
    }
  }
}
```

Results in:

```
com.apollographql.apollo.exception.ApolloParseException: Failed to parse http response
	at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:91)
	at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.access$000(ApolloParseInterceptor.java:28)
	at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor$1.onResponse(ApolloParseInterceptor.java:57)
	at com.apollographql.apollo.internal.interceptor.ApolloServerInterceptor$1$1.onResponse(ApolloServerInterceptor.java:100)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:141)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
	at java.lang.Thread.run(Thread.java:761)
Caused by: java.lang.NullPointerException: bonusFragment can't be null
	at com.ibotta.android.SearchGraphQL$Content$Fragments.<init>(SearchGraphQL.java:694)
	at com.ibotta.android.SearchGraphQL$Content$Fragments$Mapper.map(SearchGraphQL.java:799)
	at com.ibotta.android.SearchGraphQL$Content$Mapper$1.read(SearchGraphQL.java:818)
	at com.ibotta.android.SearchGraphQL$Content$Mapper$1.read(SearchGraphQL.java:815)
	at com.apollographql.apollo.internal.reader.RealResponseReader.readConditional(RealResponseReader.java:185)
	at com.ibotta.android.SearchGraphQL$Content$Mapper.map(SearchGraphQL.java:815)
	at com.ibotta.android.SearchGraphQL$Result$Mapper$1$1.read(SearchGraphQL.java:503)
	at com.ibotta.android.SearchGraphQL$Result$Mapper$1$1.read(SearchGraphQL.java:500)
	at com.apollographql.apollo.internal.reader.RealResponseReader$ListItemReader.readObject(RealResponseReader.java:252)
	at com.ibotta.android.SearchGraphQL$Result$Mapper$1.read(SearchGraphQL.java:500)
	at com.ibotta.android.SearchGraphQL$Result$Mapper$1.read(SearchGraphQL.java:497)
	at com.apollographql.apollo.internal.reader.RealResponseReader.readList(RealResponseReader.java:134)
	at com.ibotta.android.SearchGraphQL$Result$Mapper.map(SearchGraphQL.java:497)
	at com.ibotta.android.SearchGraphQL$Search$Mapper$1$1.read(SearchGraphQL.java:342)
	at com.ibotta.android.SearchGraphQL$Search$Mapper$1$1.read(SearchGraphQL.java:339)
	at com.apollographql.apollo.internal.reader.RealResponseReader$ListItemReader.readObject(RealResponseReader.java:252)
	at com.ibotta.android.SearchGraphQL$Search$Mapper$1.read(SearchGraphQL.java:339)
	at com.ibotta.android.SearchGraphQL$Search$Mapper$1.read(SearchGraphQL.java:336)
	at com.apollographql.apollo.internal.reader.RealResponseReader.readList(RealResponseReader.java:134)
	at com.ibotta.android.SearchGraphQL$Search$Mapper.map(SearchGraphQL.java:336)
	at com.ibotta.android.SearchGraphQL$Data$Mapper$1.read(SearchGraphQL.java:235)
	at com.ibotta.android.SearchGraphQL$Data$Mapper$1.read(SearchGraphQL.java:232)
	at com.apollographql.apollo.internal.reader.RealResponseReader.readObject(RealResponseReader.java:112)
	at com.ibotta.android.SearchGraphQL$Data$Mapper.map(SearchGraphQL.java:232)
	at com.ibotta.android.SearchGraphQL$Data$Mapper.map(SearchGraphQL.java:227)
	at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser$1.read(HttpResponseBodyParser.java:59)
	at com.apollographql.apollo.internal.json.ResponseJsonStreamReader.nextObject(ResponseJsonStreamReader.java:86)
	at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser.parse(HttpResponseBodyParser.java:53)
	at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:80)
	... 8 more
```

**It should be noted that the same query using inline fragments works fine. I would assume this would still work with non-inline fragments though. (I'm trying to reuse fragments across queries to reduce duplication).**

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/SchemaTypeSpecBuilder.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/SchemaTypeSpecBuilder.kt
@@ -5,11 +5,11 @@ import com.apollographql.apollo.compiler.ir.CodeGenerationContext
 import com.apollographql.apollo.compiler.ir.Field
 import com.apollographql.apollo.compiler.ir.InlineFragment
 import com.squareup.javapoet.*
-import java.io.IOException
 import javax.lang.model.element.Modifier
 
 class SchemaTypeSpecBuilder(
     val typeName: String,
+    val schemaType: String = "",
     val fields: List<Field>,
     val fragmentSpreads: List<String>,
     val inlineFragments: List<InlineFragment>,
@@ -102,7 +102,7 @@ class SchemaTypeSpecBuilder(
     fun isOptional(fragmentName: String): Boolean {
       return context.ir.fragments
           .find { it.fragmentName == fragmentName }
-          ?.let { it.typeCondition == typeName } ?: true
+          ?.let { it.typeCondition != normalizeGraphQlType(schemaType) } ?: true
     }
 
     fun fragmentFields(): List<FieldSpec> {

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ir/Field.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ir/Field.kt
@@ -24,6 +24,7 @@ data class Field(
     val fields = if (isNonScalar()) fields!! else emptyList()
     return SchemaTypeSpecBuilder(
         typeName = formatClassName(),
+        schemaType = type,
         fields = fields,
         fragmentSpreads = fragmentSpreads ?: emptyList(),
         inlineFragments = inlineFragments ?: emptyList(),

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ir/Operation.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ir/Operation.kt
@@ -15,7 +15,13 @@ data class Operation(
     val fragmentsReferenced: List<String>
 ) : CodeGenerator {
   override fun toTypeSpec(context: CodeGenerationContext): TypeSpec =
-      SchemaTypeSpecBuilder(DATA_TYPE_NAME, fields, emptyList(), emptyList(), context)
+      SchemaTypeSpecBuilder(
+          typeName = DATA_TYPE_NAME,
+          fields = fields,
+          fragmentSpreads = emptyList(),
+          inlineFragments = emptyList(),
+          context = context
+      )
           .build(Modifier.PUBLIC, Modifier.STATIC)
           .toBuilder()
           .addSuperinterface(Operation.Data::class.java)

--- a/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition/TestQuery.java
+++ b/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition/TestQuery.java
@@ -277,9 +277,9 @@ public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery
     }
 
     public static class Fragments {
-      final @Nonnull HumanDetails humanDetails;
+      final Optional<HumanDetails> humanDetails;
 
-      final @Nonnull DroidDetails droidDetails;
+      final Optional<DroidDetails> droidDetails;
 
       private volatile String $toString;
 
@@ -287,22 +287,16 @@ public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery
 
       private volatile boolean $hashCodeMemoized;
 
-      public Fragments(@Nonnull HumanDetails humanDetails, @Nonnull DroidDetails droidDetails) {
-        if (humanDetails == null) {
-          throw new NullPointerException("humanDetails can't be null");
-        }
-        this.humanDetails = humanDetails;
-        if (droidDetails == null) {
-          throw new NullPointerException("droidDetails can't be null");
-        }
-        this.droidDetails = droidDetails;
+      public Fragments(@Nullable HumanDetails humanDetails, @Nullable DroidDetails droidDetails) {
+        this.humanDetails = Optional.fromNullable(humanDetails);
+        this.droidDetails = Optional.fromNullable(droidDetails);
       }
 
-      public @Nonnull HumanDetails humanDetails() {
+      public Optional<HumanDetails> humanDetails() {
         return this.humanDetails;
       }
 
-      public @Nonnull DroidDetails droidDetails() {
+      public Optional<DroidDetails> droidDetails() {
         return this.droidDetails;
       }
 
@@ -310,11 +304,11 @@ public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery
         return new ResponseFieldMarshaller() {
           @Override
           public void marshal(ResponseWriter writer) {
-            final HumanDetails $humanDetails = humanDetails;
+            final HumanDetails $humanDetails = humanDetails.isPresent() ? humanDetails.get() : null;
             if ($humanDetails != null) {
               $humanDetails.marshaller().marshal(writer);
             }
-            final DroidDetails $droidDetails = droidDetails;
+            final DroidDetails $droidDetails = droidDetails.isPresent() ? droidDetails.get() : null;
             if ($droidDetails != null) {
               $droidDetails.marshaller().marshal(writer);
             }
@@ -482,9 +476,9 @@ public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery
     }
 
     public static class Fragments {
-      final @Nonnull HumanDetails humanDetails;
+      final Optional<HumanDetails> humanDetails;
 
-      final @Nonnull DroidDetails droidDetails;
+      final Optional<DroidDetails> droidDetails;
 
       private volatile String $toString;
 
@@ -492,22 +486,16 @@ public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery
 
       private volatile boolean $hashCodeMemoized;
 
-      public Fragments(@Nonnull HumanDetails humanDetails, @Nonnull DroidDetails droidDetails) {
-        if (humanDetails == null) {
-          throw new NullPointerException("humanDetails can't be null");
-        }
-        this.humanDetails = humanDetails;
-        if (droidDetails == null) {
-          throw new NullPointerException("droidDetails can't be null");
-        }
-        this.droidDetails = droidDetails;
+      public Fragments(@Nullable HumanDetails humanDetails, @Nullable DroidDetails droidDetails) {
+        this.humanDetails = Optional.fromNullable(humanDetails);
+        this.droidDetails = Optional.fromNullable(droidDetails);
       }
 
-      public @Nonnull HumanDetails humanDetails() {
+      public Optional<HumanDetails> humanDetails() {
         return this.humanDetails;
       }
 
-      public @Nonnull DroidDetails droidDetails() {
+      public Optional<DroidDetails> droidDetails() {
         return this.droidDetails;
       }
 
@@ -515,11 +503,11 @@ public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery
         return new ResponseFieldMarshaller() {
           @Override
           public void marshal(ResponseWriter writer) {
-            final HumanDetails $humanDetails = humanDetails;
+            final HumanDetails $humanDetails = humanDetails.isPresent() ? humanDetails.get() : null;
             if ($humanDetails != null) {
               $humanDetails.marshaller().marshal(writer);
             }
-            final DroidDetails $droidDetails = droidDetails;
+            final DroidDetails $droidDetails = droidDetails.isPresent() ? droidDetails.get() : null;
             if ($droidDetails != null) {
               $droidDetails.marshaller().marshal(writer);
             }

